File: real_benchmarks/data/real_c_rust_pairs/bitshift_overflow.c
/*
 * Bitshift overflow divergence — inspired by CVE-2015-0235 (GHOST glibc)
 * and various kernel bugs involving shift UB.
 *
 * C11 §6.5.7/3-4: shift by >= width is UB.
 * Rust: panics in debug, wraps (shift % width) in release.
 */
#include <stdint.h>

/* Shift by type width — UB in C, defined in Rust. */
uint32_t shift_by_32(uint32_t x) {
    return x << 32;  /* UB: shifting 32-bit value by 32 */
}

/* Shift by variable amount, potentially >= width. */
int32_t dynamic_shift(int32_t value, int32_t amount) {
    if (amount < 0)
        return 0;  /* negative shift is also UB in C */
    return value << amount;  /* UB if amount >= 32 */
}

/* Negative left-shift operand: UB for signed types in C. */
int32_t signed_left_shift(int32_t x, int shift) {
    return x << shift;  /* UB if x < 0 (C11 §6.5.7/4) */
}

/* Right shift of negative value: implementation-defined in C. */
int32_t arithmetic_right_shift(int32_t x, int shift) {
    return x >> shift;  /* impl-defined if x < 0: arithmetic or logical? */
}

/* Bit extraction pattern (common in crypto/protocol code). */
uint32_t extract_bits(uint32_t word, int start, int count) {
    return (word >> start) & ((1u << count) - 1);  /* UB if start >= 32 or count >= 32 */
}

File: real_benchmarks/data/real_c_rust_pairs/bitshift_overflow.rs
/*
 * Rust translation of bitshift patterns.
 * Rust: shift >= width panics in debug, wraps in release.
 * Negative shift is impossible (shift amount is unsigned or checked).
 */

/// Shift by 32: Rust wraps — 32 % 32 = 0, so result is x.
pub fn shift_by_32(x: u32) -> u32 {
    x.wrapping_shl(32)  // 32 % 32 = 0 → returns x
}

/// Dynamic shift: Rust wrapping_shl takes u32, amount is mod 32.
pub fn dynamic_shift(value: i32, amount: i32) -> i32 {
    if amount < 0 {
        return 0;
    }
    value.wrapping_shl(amount as u32)  // amount mod 32
}

/// Signed left shift: Rust wrapping_shl is defined for all values.
pub fn signed_left_shift(x: i32, shift: u32) -> i32 {
    x.wrapping_shl(shift)  // no UB, wraps
}

/// Right shift: Rust arithmetic right shift for signed types (always).
pub fn arithmetic_right_shift(x: i32, shift: u32) -> i32 {
    x.wrapping_shr(shift)  // always arithmetic for i32
}

/// Bit extraction: Rust version with wrapping shifts.
pub fn extract_bits(word: u32, start: u32, count: u32) -> u32 {
    (word.wrapping_shr(start)) & (1u32.wrapping_shl(count).wrapping_sub(1))
}
